src/Compilers/Core/Portable/InternalUtilities/BlobBuildingStream.cs

```diff
@@ -0,0 +1,126 @@
+{
+    /// <summary>
+    /// A write-only memory stream backed by a <see cref="BlobBuilder"/>.
+    /// </summary>
+    internal sealed class BlobBuildingStream : Stream
+    {
+        private static ObjectPool<BlobBuildingStream> s_pool = new ObjectPool<BlobBuildingStream>(() => new BlobBuildingStream());
+        private BlobBuilder _builder;
+
+        /// <summary>
+        /// The chunk size to be used by the underlying BlobBuilder.
+        /// </summary>
+        /// <remarks>
+        /// The current single use case for this type is embedded sources in PDBs.
+        ///
+        /// 32 KB is:
+        ///
+        /// * Large enough to handle 99.6% all VB and C# files in Roslyn and CoreFX 
+        ///   without allocating additional chunks.
+        ///
+        /// * Small enough to avoid the large object heap.
+        ///
+        /// * Large enough to handle the files in the 0.4% case without allocating tons
+        ///   of small chunks. Very large source files are often generated in build
+        ///   (e.g. Syntax.xml.Generated.vb is 390KB compressed!) and those are actually
+        ///   attractive candidates for embedding, so we don't want to discount the large
+        ///   case too heavily.)
+        ///
+        /// * We pool the outer BlobBuildingStream but only retain the first allocated chunk.
+        /// </remarks>
+        public const int ChunkSize = 32 * 1024;
+
+        public override bool CanWrite => true;
+        public override bool CanRead => false;
+        public override bool CanSeek => false;
+        public override long Length => _builder.Count;
+
+        public static BlobBuildingStream GetInstance()
+        {
+            return s_pool.Allocate();
+        }
+
+        private BlobBuildingStream()
+        {
+            // NOTE: We pool the wrapping BlobBuildingStream, but not individual chunks.
+            // The first chunk will be reused, but any further chunks will be freed when we're done building blob.
+            _builder = new BlobBuilder(ChunkSize);
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            _builder.WriteBytes(buffer, offset, count);
+        }
+
+        public override void WriteByte(byte value)
+        {
+            _builder.WriteByte(value);
+        }
+
+        public void WriteInt32(int value)
+        {
+            _builder.WriteInt32(value);
+        }
+
+        public void PatchLeadingInt32(int value)
+        {
+            foreach (var blob in _builder.GetBlobs())
+            {
+                ArraySegment<byte> segment = blob.GetBytes();
+                Debug.Assert(segment.Count > 4);
+
+                byte[] array = segment.Array;
+                array[0] = (byte)(value & 0xFF);
+                array[1] = (byte)((value >> 8) & 0xFF);
+                array[2] = (byte)((value >> 16) & 0xFF);
+                array[3] = (byte)((value >> 24) & 0xFF); 
+
+                return;
+            }
```

Comment:
Seems too complicated. Couldn't you reserve 4 bytes at the beginning (BlobBuilder.ReserveBytes) and then simply write to the reserved bytes?